flask/app.py

```diff
@@ -929,7 +929,7 @@ def run(self, host=None, port=None, debug=None,
             sn_host, _, sn_port = server_name.partition(':')
 
         host = host or sn_host or _host
-        port = int(port or sn_port or _port)
+        port = int(next((p for p in (port, sn_port) if p is not None), _port))
```

Comment:
this is kinda cryptic and hard to understand. I'd suggest
`int(next(p for p in (port, sn_port, _port) if p is not None))`
I guess that also allows `sn_port` to be 0, so if you use that, you should add a test for that. Alternatively you could specifically allow `port` alone to be 0 using:
`int(port) if port is not None else int(sn_port or _port)`

Reply:
Hi @ocamlmycaml, sorry for the long delay.

I added a test for `sn_port` being zero in https://github.com/pallets/flask/pull/2928/commits/583da58419e82adc672b5565587d143fbf5b66a6

regarding readability, I kinda agree with you:

```
int(next(p for p in (port, sn_port, _port) if p is not None))
```

is shorter and more intuitive than

```
int(next((p for p in (port, sn_port) if p is not None), _port))
```

I'm having a little problem with your suggestion because it changes the behavior of original code, checking also if `_port` is not None.

How about I rename `_host` and `_port` to `_default_*`? That would make it 

```
int(next((p for p in (port, sn_port) if p is not None), _default_port))
```

which IMO shows the purpose a little better.